### PR TITLE
ARROW-3738: [C++] Parse ISO8601-like timestamps in CSV columns

### PR DESCRIPTION
**Pull request by pitrou**
Second granularity is allowed (we might want to add support for fractions of seconds, e.g. in the "YYYY-MM-DD[T ]hh:mm:ss.ssssss" format).

Timestamp conversion also participates in CSV type inference, since it's unlikely to produce false positives (e.g. a semantically "string" column that would be entirely made of valid timestamp strings).

### CODE REVIEW

File: cpp/src/arrow/util/date.h
Comment (pitrou) [inline cpp/src/arrow/util/date.h]: Looks like `clang-format` ran over this file...

File: python/pyarrow/tests/test_csv.py
Comment (xhochy) [inline python/pyarrow/tests/test_csv.py]: Shouldn't this be a `date` column and only a `datetime` column when it includes hours/minutes?

Comment (xhochy) [reply python/pyarrow/tests/test_csv.py]: What about adding a time to these so that we don't have a test that would break when we add `date` support?

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: Well, we can fix the test by then. Right now those are inferred as timestamps, and that's what the test checks for.

**Comment by pitrou** [comment]
Will merge soon if no CI fail.

**Comment by wesm** [comment]
I'd still like to have a look -- let me have a look this morning and will merge if no issues

**Comment by wesm** [comment]
Sorry to be a bit delayed -- week of Thanksgiving in the US in always a bit challenging. I will try to rebase this and give it a quick review before merging

File: cpp/src/arrow/util/parsing.h
Comment (wesm) [inline cpp/src/arrow/util/parsing.h]: Do you have a sense of performance of this?

**Review by wesm — APPROVED** [review]
+1. I rebased, so merge this once the build is passing again

File: cpp/src/arrow/util/parsing.h
Comment (pitrou) [reply cpp/src/arrow/util/parsing.h]: No idea. Hopefully it shouldn't be too slow...

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-3853 about adding a cast implementation that uses this. We should also add benchmarks once we do that

**Pull request closed without merging**
